xpcom/io/nsPipe3.cpp

```diff
@@ -1660,10 +1660,15 @@ nsPipeOutputStream::AddRef() {
 NS_IMETHODIMP_(MozExternalRefCountType)
 nsPipeOutputStream::Release() {
   if (--mWriterRefCnt == 0) {
-    // Because the refcount is threadsafe, the final release can occur at
-    // non-deterministic points.
-    recordreplay::AutoDisallowThreadEvents disallow;
-    Close();
+    // Because the refcount is threadsafe and can be held by GC'ed objects,
+    // the final release can occur at non-deterministic points. Closing the
+    // output stream affects the pipe's state and consequently the behavior
+    // of input streams that may still be in use. To ensure the input streams
+    // behave the same when replaying we don't close the output stream after
+    // the last release when recording/replaying.
+    if (!recordreplay::IsRecordingOrReplaying()) {
+      Close();
```

Comment:
Could this close be deterministically scheduld for close?  Wouldn't this affect page behaviour in the sense of breaking expected JS stream semantics by code?

Reply:
This shouldn't break JS as the close behavior here is non-deterministic anyways and is more internal book-keeping by the browser.  I don't think it will have a big impact as we're not doing anything to keep the pipe itself alive and when the last references on the input streams have been released then everything will be torn down.

We used to have a mechanism for triggering events like this at deterministic points, but it was complicated and could get the browser into states where its invariants were broken and we'd get crashes that were very hard to debug.  That's all been ripped out now in favor of letting resources be torn down non-deterministically during GC, and leaking references (like we're doing here) when that causes problems.  This has been a big improvement to stability, simplicity, and performance and I'd rather not add back something like the old mechanism.